codex/stores/queryiterhelper.nim

```diff
@@ -0,0 +1,66 @@
+import pkg/questionable
+import pkg/questionable/results
+import pkg/chronos
+import pkg/chronicles
+import pkg/datastore/typedds
+
+import ../utils/asynciter
+
+type KeyVal[T] = tuple[key: Key, value: T]
+
+proc toAsyncIter*[T](
+  queryIter: QueryIter[T],
+  finishOnErr: bool = true
+  ): Future[?!AsyncIter[?!QueryResponse[T]]] {.async.} =
+  ## Converts `QueryIter[T]` to `AsyncIter[?!QueryResponse[T]]` and automatically
+  ## runs dispose whenever `QueryIter` finishes or whenever an error occurs (only
+  ## if the flag finishOnErr is set to true)
```

Comment:
There is another case in which a queryIter must be disposed. This is when an iteration is stopped early, before all items have been yielded. LevelDB showed that we were leaking unfinished iters. Sqlite would clean them up automatically on db.close. But LevelDB will throw instead.

I'm not sure this impacts your changes. But please make sure that the QueryIter is disposed correctly, even when the async iter does not error and does not run to completion.

See the TODO here `reservations.nim` in `proc findAvailability`: https://github.com/codex-storage/nim-codex/blob/3246c43174fd5c3be9ad422d6c35431cccafc6c0/codex/sales/reservations.nim#L629

Reply:
Thanks for this comment. I realise that there's a flow like that, but I didn't realise we can address this flow when we have ARC. Current use cases are not finishing iteration prematurely, so either all items have been traversed or an error occurred - both cases are handled.